donut/modules/marketplace/helpers.py

```diff
@@ -0,0 +1,274 @@
+import flask
+import sqlalchemy
+from datetime import date, datetime
+
+import routes
+
+
+def render_top_marketplace_bar(template_url, **kwargs):
+    """
+    Provides an easy way for routing functions to pass the variables required for
+    rendering the marketplace's top bar to render_template.  Basically chains
+    some other arguments on to the render call, namely the list of marketplace
+    categories, the list urls to categories, and the width of each url column.
+
+    Arguments:
+        template_url: The url which is being rendered.
+        **kwargs: The variables which are used to render the rest of the page.
+
+    Returns:
+        The result of render_template(): Whatever magic Flask does to render the
+                                         final page.
+    """
+    query = sqlalchemy.sql.select(["cat_title"]).select_from(sqlalchemy.text("marketplace_categories"))
+    result = flask.g.db.execute(query)
+
+    # Convert result, in the form of a strange SQL object, into an array,
+    # so that we can retrieve the length without having to execute a special
+    # length query or anything.
+    categories = []
+    for column in result:
+        categories.append(column[0])
+
+    # Get number of rows and columns to display categories nicely. This is a
+    # little tricky - we want to aim for a table with either 4 categories or 5
+    # categories per row, with the last row catching the remainder. However, we
+    # want to avoid having exactly 1 category in the remainder row if possible.
+    num_cats = len(categories);
+    num_cols = num_cats
+    if num_cats <= 5:
```

Comment:
This logic is a bit hard to understand. Is it completely necessary or is there a different way of 
displaying? I'm also not completely sure what the issue is. Maybe a drawing would help.